csharp/src/Ice/AsyncSemaphore.cs

```diff
@@ -46,31 +63,42 @@ internal async ValueTask WaitAsync(CancellationToken cancel = default)
             }
             catch (Exception ex)
             {
-                taskCompletionSource.TrySetException(ex);
+                lock (_mutex)
+                {
+                    if (_queue.Contains(taskCompletionSource))
+                    {
+                        taskCompletionSource.SetException(ex);
+                    }
+                }
                 throw;
             }
         }
 
         internal void Release()
         {
-            ManualResetValueTaskCompletionSource<bool> taskCompletionSource;
-            do
+            lock (_mutex)
             {
-                lock (_mutex)
+                if (_currentCount == _maxCount)
                 {
-                    if (_queue.Count == 0)
+                    throw new SemaphoreFullException($"semaphore maximum count of {_maxCount} already reached");
```

Comment:
While this can't happen now we could have code where this could be useful. It's also easier to debug bugs with this exception (the reason why I added it with this change).